packages/central-server/src/apiV2/utilities/permissionsBasedChangeFilter.js

```diff
@@ -0,0 +1,170 @@
+
+export const supportsPermissionsBasedSync = version =>
+  semverCompare(version, PERMISSIONS_BASED_SYNC_MIN_APP_VERSION) >= 0;
+
+export const getCountriesAndPermissionsToSync = async req => {
+  const { accessPolicy } = req;
+  const {
+    appVersion,
+    countriesSynced: countriesSyncedString,
+    permissionGroupsSynced: permissionGroupsSyncedString,
+  } = req.query;
+
+  if (!supportsPermissionsBasedSync(appVersion)) {
+    throw new Error(
+      `Permissions based sync is not supported for appVersion: ${appVersion}, must be ${PERMISSIONS_BASED_SYNC_MIN_APP_VERSION} or higher`,
+    );
+  }
+
+  const usersCountries = accessPolicy.getEntitiesAllowed();
+  const usersPermissionGroups = accessPolicy.getPermissionGroups();
+
+  // First time sync, just return new countries and permission groups
+  if (!countriesSyncedString || !permissionGroupsSyncedString) {
+    const userCountryIds = (await req.models.country.find({ code: usersCountries })).map(
+      country => country.id,
+    );
+    return {
+      unsynced: {
+        countries: usersCountries,
+        countryIds: userCountryIds,
+        permissionGroups: usersPermissionGroups,
+      },
+    };
+  }
+
+  const syncedCountries = countriesSyncedString.split(',');
+  const syncedPermissionGroups = permissionGroupsSyncedString.split(',');
+  const syncedCountryIds = (await req.models.country.find({ code: syncedCountries })).map(
+    country => country.id,
+  );
+
+  const unsyncedCountries = usersCountries.filter(country => !syncedCountries.includes(country));
+  const unsyncedPermissionGroups = usersPermissionGroups.filter(
+    permissionGroup => !syncedPermissionGroups.includes(permissionGroup),
+  );
+
+  if (unsyncedCountries.length === 0 && unsyncedPermissionGroups.length === 0) {
+    // No new countries or permissionGroups, just return synced
+    return {
+      synced: {
+        countries: syncedCountries,
+        countryIds: syncedCountryIds,
+        permissionGroups: syncedPermissionGroups,
+      },
+    };
+  }
+
+  const unsyncedCountryIds = (await req.models.country.find({ code: unsyncedCountries })).map(
+    country => country.id,
+  );
+
+  return {
+    unsynced: {
+      countries: unsyncedCountries,
+      countryIds: unsyncedCountryIds,
+      permissionGroups: unsyncedPermissionGroups,
+    },
+    synced: {
+      countries: syncedCountries,
+      countryIds: syncedCountryIds,
+      permissionGroups: syncedPermissionGroups,
+    },
+  };
+};
+
+export const getPermissionsWhereClause = permissionsBasedFilter => {
+  let query = '';
+  const params = [];
+  const { countries, countryIds, permissionGroups } = permissionsBasedFilter;
+  const typesWithoutPermissions = ['country', 'permission_group']; // Sync all countries and permission groups (needed for requesting access)
+  const permissionsClauses = [
+    {
+      query: `"type" = ? AND record_type IN ${SqlQuery.record(typesWithoutPermissions)}`,
```

Comment:
I'd prefer to separate these into two separate clauses, as it confused me initially thinking that we were only excluding deletes for the record types without permissions. I understand that is hard with the `OR`s between everything, so maybe we set up an initial query string outside of this first?

Reply:
I think I can come up with something to address this 👍